### ISSUE TITLE
@babel-register is not configurable when used with ES Modules (--experimental-modules)

### ISSUE THREAD

**Issue by feross** [babel/babel#6737]
Choose one: is this a bug report or feature request? Bug report

### Input Code

```js
import babelRegister from '@babel/register'
babelRegister({ extensions: '.xxx' })
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

The specific config I'm using should not be relevant, but I'm providing it just in case.

```js
{
    "plugins": [
      "transform-class-properties",
      [
        "transform-react-jsx",
        {
          "pragma": "jsx",
          "useBuiltIns": true
        }
      ],
      "transform-object-rest-spread"
    ]
  }
```

### Expected Behavior

Invoking Node.js 8 LTS or Node.js 9 with the `--experimental-modules` command line flag enables built-in support for ES Modules. When this option is passed, the entry-point file must be an ES Module, which requires the use of `import` statements; `require` is not available. Using `import` on CommonJS modules is supported.

I expect that invoking the following code will change the babel-register configuration:

```js
import babelRegister from '@babel/register'
babelRegister({ extensions: '.xxx' })
```

### Current Behavior

The configuration stays unchanged even when following how the documentation says to configure `babel-register`.

### Possible Solution

I think the reason might be that module loading with ES Modules is async, while CommonJS is sync. Does `babel-register` assume that the function to configure will be called in the first tick?

### Context

Trying to using `babel-register` for server-side JSX support, while also using node's built-in ES Module support with the `--experimental-modules` flag.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  @babel/core 7.0.0-beta.4; @babel-register 7.0.0-beta.5
| Babylon          | 
| node             | 8.9.0
| npm              | 5.5.1
| Operating System | macOS 10.13.1

**Comment by loganfsmyth** [comment]
I don't believe Node's ES6 `import` statements will run require hooks, so I don't think there is any way for babel to hook into those, at least currently.

**Comment by feross** [comment]
@loganfsmyth Require hooks are getting run when using `import` on a CommonJS module (a file with a .js extension).

`babel-register` does run on this import:

```js
import app = './app' // this references a commonjs file called app.js
```

**Comment by Kovensky** [comment]
NOTE: there is no guarantee that this is how it is _intended_ to work. It is possible that `@babel/register` works on `import`-ed commonjs by accident instead of by design.

When something is `import`ed (statically or dynamically), it goes through the `ModuleRequest` loader. The loader responsible for loading `.js` files, aka [the `cjs` loader](https://github.com/nodejs/node/blob/master/lib/internal/loader/ModuleRequest.js#L37-L45), just delegates it directly to the current commonjs loader, which happens to invoke handlers in `require.extensions`.

However, there is a [probably incomplete API for hooking into ESM imports as well](https://github.com/nodejs/node/blob/master/lib/module.js#L456-L461), which may be usable in the future; but probably not via `@babel/register`.

**Comment by graingert** [comment]
The `--loader` hook has now been documented: https://nodejs.org/api/esm.html#esm_loader_hooks

You can set it via --loader eg `@babel/register` could provide a `@babel/register/index.mjs` for using the node loader hooks.

**Comment by pruett** [comment]
Could someone provide the full answer to this issue? Running into something similar. `--loader` doesn't seem to help

**Comment by graingert** [comment]
@pruett it's not implemented yet

**Comment by pruett** [comment]
ah, thanks @graingert !
